### PR TITLE
feat: created endpoints for user functionality

### PR DESCRIPTION
**Pull request by Ethel3**

### CODE REVIEW

File: controllers/user.controller.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// controllers/userController.mjs
```
Comment (kwakubiney) [inline controllers/user.controller.js]: Let's remove all comments like this

File: controllers/user.controller.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import pool from '../dbConfig.mjs';
+
+export const createUser = async (req, res) => {
+  try {
+    const { username, age, email } = req.body;
+    const query = `
+      INSERT INTO users (username, age, email)
+      VALUES ($1, $2, $3)
+      RETURNING *;
```
Comment (kwakubiney) [inline controllers/user.controller.js:9]: Wow. Never used `RETURNING *` before. 
TIL.

File: controllers/user.controller.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    `;
+    const values = [username, age, email];
+    const result = await pool.query(query, values);
+    res.json(result.rows[0]);
+  } catch (error) {
+    res.status(500).json({ error: 'An error occurred.' });
```
Comment (kwakubiney) [inline controllers/user.controller.js]: Errors should be descriptive enough for the client. Maybe we can give much more information in the error message? 
```suggestion
    res.status(500).json({ error: 'Failed to create new user.' });
```
And also possibly log the error with a `console.error()` before returning the json so we can tell what the problem was

File: controllers/user.controller.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    const { id } = req.params;
+    const { username, age, email } = req.body;
+    const query = `
+      UPDATE users
+      SET username = $1, age = $2, email = $3
+      WHERE id = $4
+      RETURNING *;
+    `;
+    const values = [username, age, email, id];
+    const result = await pool.query(query, values);
```
Comment (kwakubiney) [inline controllers/user.controller.js]: There is an edge case you're missing here.
What if no user exists with a `userID` as the one you passed in the `params`? 

Hint: You have to check if such a user exists with that `ID` before proceeding to make the update query. Appropriate status codes must be used if user is not found :)

File: controllers/user.controller.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  try {
+    const { username, age, email } = req.body;
+    const query = `
+      INSERT INTO users (username, age, email)
+      VALUES ($1, $2, $3)
+      RETURNING *;
+    `;
+    const values = [username, age, email];
+    const result = await pool.query(query, values);
+    res.json(result.rows[0]);
```
Comment (kwakubiney) [inline controllers/user.controller.js]: It's advised to include a status code even if it's successful, with an appropriate response body. Let's work with uniformity here.
```suggestion
    res.status(<appropriate status code>).json({
    "message":"User created successfully",
    "data": result.rows[0]});
```
We can apply the same structure to all successful responses.

`TODO`: Look for appropriate status code to be used when a resource is successfully created on the backend.

File: controllers/user.controller.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      SET username = $1, age = $2, email = $3
+      WHERE id = $4
+      RETURNING *;
+    `;
+    const values = [username, age, email, id];
+    const result = await pool.query(query, values);
+    res.json(result.rows[0]);
+  } catch (error) {
+    res.status(500).json({ error: 'An error occurred.' });
+  }
```
Comment (kwakubiney) [inline controllers/user.controller.js]: Ditto

File: controllers/user.controller.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+};
+
+export const deleteUser = async (req, res) => {
+  try {
+    const { id } = req.params;
+    const query = `
+      DELETE FROM users
+      WHERE id = $1;
+    `;
+    const values = [id];
```
Comment (kwakubiney) [inline controllers/user.controller.js]: Same thing. Check if user exists before

File: controllers/user.controller.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    await pool.query(query, values);
+    res.json({ message: 'User deleted successfully.' });
+  } catch (error) {
+    res.status(500).json({ error: 'An error occurred.' });
+  }
+};
+
+export const listUsers = async (req, res) => {
```
Comment (kwakubiney) [inline controllers/user.controller.js]: Consider renaming to `getUsers`

File: controllers/user.controller.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    const { id } = req.params;
+    const query = `
+      DELETE FROM users
+      WHERE id = $1;
+    `;
+    const values = [id];
+    await pool.query(query, values);
+    res.json({ message: 'User deleted successfully.' });
+  } catch (error) {
+    res.status(500).json({ error: 'An error occurred.' });
```
Comment (kwakubiney) [inline controllers/user.controller.js]: More descriptive errors please

File: controllers/user.controller.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+export const listUsers = async (req, res) => {
+  try {
+    const query = `
+      SELECT * FROM users;
+    `;
+    const result = await pool.query(query);
+    res.json(result.rows);
+  } catch (error) {
+    res.status(500).json({ error: 'An error occurred.' });
```
Comment (kwakubiney) [inline controllers/user.controller.js]: Ditto

File: routes/user.routes.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+// routes/userRoutes.mjs
```
Comment (kwakubiney) [inline routes/user.routes.js]: Remove comments

File: middleware/errorHandler.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+const errorHandler = (err, req, res)=>{
+    const statusCode = res.statusCode === 200 ? 500 : res.statusCode;
+    res.status(statusCode);
+    res.json({
+        status: 'fail',
+        message: err.message,
+        stack: process.env.NODE_ENV === 'production' ? null : err.stack,
+    });
+}
```
Comment (kwakubiney) [inline middleware/errorHandler.js]: Oh just saw this error handler. I think there's a bug in your logic though. Investigate `const statusCode = res.statusCode === 200 ? 500 : res.statusCode;`

I am assuming this is supposed to handle `500` errors?

File: controllers/user.controller.js (hunk 6 above)
Comment (Ethel3) [reply controllers/user.controller.js]: I do not understand this

Comment (kwakubiney) [reply controllers/user.controller.js]: Just means "Applicable to what I mentioned before"

File: controllers/user.controller.js [hunk 13]
```diff
@@ -1,8 +1,8 @@
-// controllers/userController.mjs
 import pool from '../dbConfig.mjs';
 
-export const createUser = async (req, res) => {
+export const createUser = async (req, res, next) => {
   try {
+    console.log(req.body);
```
Comment (kwakubiney) [inline controllers/user.controller.js]: Let's remove this `console.log()`.

Comment (Ethel3) [reply controllers/user.controller.js]: including the try?

Comment (kwakubiney) [reply controllers/user.controller.js]: No, just the console log

File: dbConfig.mjs [hunk 14]
```diff
@@ -1,17 +1,14 @@
...
+dotenv.config()
+const pool = new Pool({
+    user: process.env.DB_USERNAME,
+    host: process.env.DB_HOST,
+    database: process.env.DB_DATABASE_NAME,
+    password: process.env.DB_PASSWORD,
+    port: process.env.DB_PORT,
+})
+
+export default pool
```
Comment (kwakubiney) [inline dbConfig.mjs]: Any reason for changing this and using `pool` instead of using the `connectToDb()` function?

File: index.mjs [hunk 15]
```diff
@@ -1,13 +1,19 @@
 import express from 'express';
 import bodyParser from 'body-parser';
 import { createRequire } from 'module';
-import connectToDb from './dbConfig.mjs';
+// import connectToDb from './dbConfig.mjs';
```
Comment (kwakubiney) [inline index.mjs:4]: Why is this commented out?

File: index.mjs [hunk 16]
```diff
@@ -1,13 +1,19 @@
...
+import errorHandler from './middleware/errorHandler.js';
+import router from './routes/user.routes.js';
 
 const require = createRequire(import.meta.url)
-const app = express();
 require('dotenv').config()
-connectToDb()
+// connectToDb()
+const app = express();
```
Comment (kwakubiney) [inline index.mjs:11]: This too?

File: dbConfig.mjs (hunk 14 above)
Comment (Ethel3) [reply dbConfig.mjs]: Yes. It allows multiple users to access the database at the same time and helps to manage and reuse connections more efficiently.

Comment (Ethel3) [reply dbConfig.mjs]: It also binds the queries to the available connections.

Comment (kwakubiney) [reply dbConfig.mjs]: I agree a `pool` does that but this change does not really differ from what we did earlier. We are still using a `pool` but just calling `connect` in a function so we can just export the function

Comment (kwakubiney) [reply dbConfig.mjs]: Also, it's advisable to keep your config stuff inside the entry point of your application, which is `index.js`. Consider moving `dotenv.config()` back to where it was

Comment (kwakubiney) [reply dbConfig.mjs]: >I agree a pool does that but this change does not really differ from what we did earlier. We are still using a pool but just calling connect in a function so we can just export the function

On second thought, I might be wrong judging from the docs. You can just use a `pool` object to query for stuff without necessarily connecting. I will have to read on that but we can go with your change for now

Comment (kwakubiney) [reply dbConfig.mjs]: > It also binds the queries to the available connections.

Good catch

**Pull request closed without merging**